### ISSUE TITLE
JoSS Review: Using Virtual Scanner

### ISSUE THREAD

**Issue by vsoch** [imr-framework/virtual-scanner#36]
This is for https://github.com/openjournals/joss-reviews/issues/1637 and note that I'm testing with the container from #35. These are some issues related to using the software.

## Login
As a user, I'm tempted to just select standard or advanced and click "Start Scan" but nothing happens. I want to suggest any of the following:

 - Don't require an email to use it (this seems like an unnecessary bit of personal information for a user to provide to a development server).
 - If the email must be required (why?) disable the button until it's entered.
 - if the email isn't entered and I click start scan, I should see the box highlighted in red and/or a notification that I need to enter my email before continuing.

We want to avoid the current setup where I can click the button and get no response from the interface about anything being wrong.

## Subject Entering
When I first log in, it's very intuitive to enter a subject (I made a bunch of fields up) on the Register screen. I then proceeded to the Acquisition. However, when I selected the sequence and then clicked "Run Scan" it jumps me back to the register page and tells me I need to register a subject. I "think" what it needs me to do is select "Numerical" from the dropdown (register a phantom?) and then continue, but it's not obvious from this usage. I'd like to suggest that if I need to register a phantom (or numerical) it takes me back to that page, with that specific error message, and the Numerical tab selected instead of Subject.

## Run Scan

Once I entered a numerical subject, I clicked Run Scan and there was an error (and here is the traceback - an index out of bounds):

```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/opt/conda/lib/python3.7/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/opt/conda/lib/python3.7/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "/opt/conda/lib/python3.7/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/opt/conda/lib/python3.7/site-packages/virtual_scanner-1.0.21-py3.7.egg/virtualscanner/coms/coms_ui/coms_server_flask.py", line 339, in worker
    progress = bsim.run_blochsim(seqinfo=payload, phtinfo=rows[0][0],
IndexError: list index out of range
172.17.0.1 - - [24/Aug/2019 19:56:21] "GET /receiver?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 200 -
```
Likely this is because some input parameter was incorrect / missing, so my suggestion is to do validation of all parameters first, and then return to the form if anything is invalid. In fact, this shouldn't happen only on the Run Scan page, but also on the Register page - I shouldn't be allowed to register with missing / incorrect data, or to click Run Scan without having a validation first.

I'll continue my review after these are updated - feel free to merge #35 so you can start with the updated base (and make adding these next changes easier, without merge conflicts, etc.)

**Comment by tonggehua** [comment]
@vsoch they should be resolved now with cb802265e0e337eec3deff3e7c2ec3faa0b4685e

**Comment by vsoch** [comment]
Thanks! I'm pretty busy atm so I'll need to pick this up later. It's "kept as new" in my inbox :)

**Comment by sairamgeethanath** [comment]
Given that we are unable to dockerize the code and reproduce errors from docker in the local environment, should we do the docker for the next version of the release? @tonggehua @sravan953 @Mmj94 @vsoch @mathieuboudreau @qianenlin

**Comment by mathieuboudreau** [comment]
I'm satisfied with at least having a virtual environment setup, I'm not completely attached to having a Docker one in this release.

**Comment by vsoch** [comment]
I didn't have any issues with the Docker recipe I produced... Dockerization is a requirement on my end.

**Comment by vsoch** [comment]
Virtual environments are one step up from not providing anything, but virtual environments != reproduciblity. I [mentioned this requirement](https://github.com/openjournals/joss-reviews/issues/1590#issuecomment-515681126) when I agreed to review, so I hope it is not coming as a surprise. I can help in whatever way is necessary, perhaps you can start by sharing what is different about the container and running locally?

**Comment by tonggehua** [comment]
We are not sure what is different between the container and running locally - not enough to resolve the bug. The crash mentioned earlier in https://github.com/imr-framework/virtual-scanner/issues/36#issue-484857122 and also a similar crash on Analyze page both happen only when we use Docker, not when we are running within the local virtual environment, with Pycharm. We've tried to debug with Docker and Pycharm but ran into problems. In addition, the supported version of Docker doesn't run on Windows 10 Home which is what two of our lab mates use, and likely many of our potential users. We are open to suggestions on how to debug :)

**Comment by vsoch** [comment]
Windows still has a lot of issues with using Docker - do you have a linux machine to test it on? Or even a Mac? Or can you grab a cloud instance and run there, exposing the port to your local machine?

The target audience is usually not a Windows user, as containers and windows are a relatively (still) buggy thing, but that of course is my opinion. I would imagine someone on linux wanting to run it (or more likely) wanting to deploy it for a lab. [Most deployment servers are linux](https://www.zdnet.com/article/linux-foundation-finds-enterprise-linux-growing-at-windows-expense/).

As a sidenote, if your entire lab is truly running windows (which makes sense, I came from a neuropsych lab and everyone was using Windows and Matlab) I would strongly consider, even for a small personal machine, having linux (ubuntu flavor or similar even) to use for programming and development. My learning increased exponentially in graduate school when my (previous) windows machine crashed, and I really was forced to work on a command line, compile things, etc.

**Comment by tonggehua** [comment]
Crash errors and Docker build problems were fixed with 6a19c16d4496dd036e94f077c6584c0ad735e700, with a new Dockerfile update. Right now, Windows users should only use the published image (there is an OSError when trying to run a locally built image) and Mac users can either use the published image or build locally (instructions in README.md have been updated). We currently do not have a linux setup but would love to look into it if needed. Please review. Again, thanks for the suggestions!

**Comment by vsoch** [comment]
Can you please explain what you mean by not having a Linux set up? Containers run native we on Linux, much more easily than Mac or Windows.  Original recipe that I developed ran cleanly on Ubuntu.

**Comment by vsoch** [comment]
If you just need a test of the new recipe on Linux I’d be happy to do this tomorrow :)

**Comment by vsoch** [comment]
oy vey, my work wireless is still out (and I'm teathered to my phone) so I'll need to wait to test this for when I go back to my apartment later this afternoon. Even cloning a git repo is substantially slow / takes a lot of bandwidth!

**Comment by tonggehua** [comment]
@vsoch Thanks so much for the response! By "not having a Linux setup", I mean that we don't a Linux system installed and in use for research on lab/personal computers.

**Comment by vsoch** [comment]
ah okay! I thought you meant that the container wouldn't work for linux (it should!) I'll let you know when I'm able to test.

**Comment by vsoch** [comment]
heyo! I didn't get to test it today, but I made sure to find a spot with a good connection, clone the repo, and build the container - so I should be able to complete my test and review tomorrow morning.

**Comment by vsoch** [comment]
okay testing! The start up went much more smoothly this time - I could click a mode and proceed without worrying about email. The subject was then largely set up and I just needed to continue to analyze, and then the analyze (albeit a bit slow, I know you'll work on it) did produce an Axial image. I tested both GE and SE, and also the Analyze tab with the sample data. I'm not totally clear on what this is useful for, but I'm not actively working in this area so I don't expect to have this. 

For the future - it could be that the slowness might be helped via using some sort of multiprocessing, or conversion to a Singularity container that can better leverage resources on the host (e.g., gpu libraries). My last suggestion is to put somewhere a note about expected processing times, because they are upwards of minutes, at least testing in the Docker container.

Otherwise, great job on the fixes here! I'm looking forward to seeing how this is further developed.
